pkg/parameters/aws/aws.go

```diff
@@ -0,0 +1,213 @@
+		klog.Warningf("trying to autodetect AWS instance ID: %v", err)
+		return ""
+	}
+
+	klog.V(2).Infof("detected AWS instance ID: %q", instanceID)
+	return instanceID
+}
+
+func getAWSConfig() *aws.Config {
+	region := os.Getenv("AWS_REGION")
+	if region == "" {
+		region = os.Getenv("AWS_DEFAULT_REGION")
+	}
+	if region == "" {
+		region = detectRegion()
+	}
+
+	httpClient := &http.Client{
+		Timeout: awsTimeout,
+	}
+	config := aws.NewConfig().WithHTTPClient(httpClient).WithRegion(region)
+	return config
+}
+
+func NewAWSParameters(base string, config *aws.Config) (*AWSParameters, error) {
+	if config == nil {
+		config = getAWSConfig()
+	}
+
+	instanceID := detectInstanceID()
+	if instanceID == "" {
+		return nil, fmt.Errorf("failed to detect AWS instance ID, not running on AWS?")
+	}
+
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+	ec2Client := ec2.New(sess, config)
+
+	sess, err = session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+	ssmClient := ssm.New(sess, config)
+
+	return &AWSParameters{
+		base: filepath.Join(base, instanceID),
+		ec2:  ec2Client,
+		ssm:  ssmClient,
+	}, nil
+}
+
+func unmarshalParameters(params map[string]string) (map[string]string, error) {
+	// This is either one single SSM parameter if it fit (<=4096 bytes), or
+	// chunks of the serialized config e.g. "config-0", "config-1", etc. We
+	// need to assemble the chunks in order, de-serialize and re-create the
+	// original config map.
+	if len(params) == 0 {
+		return nil, fmt.Errorf("got no SSM parameters")
+	}
+
+	// The whole map in one parameter.
+	if len(params) == 1 {
+		for k, v := range params {
+			configMap := make(map[string]string)
+			klog.V(2).Infof("unmarshaling SSM parameter %q", k)
+			err := yaml.Unmarshal([]byte(v), &configMap)
+			if err != nil {
+				return nil, err
+			}
+			return configMap, nil
+		}
+	}
+
+	// Multiple chunks.
+	paramList := make([]string, len(params))
+	for k, v := range params {
+		parts := strings.SplitN(k, "-", 2)
+		if len(parts) < 2 {
+			return nil, fmt.Errorf("invalid SSM parameter chunk key: %s", k)
+		}
+		n, err := strconv.ParseUint(parts[1], 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		if int(n) >= len(paramList) {
+			return nil, fmt.Errorf("invalid SSM parameter chunk key: %s", k)
+		}
+		paramList[n] = v
+	}
+	yml := ""
+	for i := range paramList {
+		yml = yml + paramList[i]
+	}
+	configMap := make(map[string]string)
+	klog.V(2).Infof("unmarshaling %d SSM parameter chunks, %d bytes", len(paramList), len(yml))
+	err := yaml.Unmarshal([]byte(yml), &configMap)
+	if err != nil {
+		return nil, err
+	}
+	return configMap, nil
+}
+
+func (a *AWSParameters) getParameter(name string) string {
+	path := filepath.Join(a.base, name)
+	in := &ssm.GetParameterInput{
+		Name:           aws.String(path),
+		WithDecryption: aws.Bool(true),
+	}
+	out, err := a.ssm.GetParameter(in)
+	if err != nil {
+		klog.V(2).Infof("getting SSM parameter %s: %v", name, err)
```

Comment:
Nitpicking: from function doc string I see that they recommend casting returned error to `awserr.Error` to get more info:
```
// Returns awserr.Error for service API and SDK errors. Use runtime type assertions
// with awserr.Error's Code and Message methods to get detailed information about
// the error.
```
I don't know if it's needed, your call ;)

Reply:
I'd say that's useful if we would like to act based on what kind of error happened (e.g. a rate limiting error from the API means we should retry after waiting). Here I kept it simple, and failures like this one will lead to itzo-launcher exiting after the error propagated back to top level, letting systemd restart it.